project.php

```diff
@@ -1313,15 +1314,18 @@ function do_history()
 
 /**
  * If the project's event-history has gaps, fill them with pseudo-events.
+ *
+ * @return array{'timestamp':int, 'who':string, 'event_type':string, 'details1':string, 'details2':string, 'details3':string}[]
  */
-function fill_gaps_in_events($in_events)
+// TODO(jchaffraix): Add a class for ProjectEvent and switch to this function to it.
+function fill_gaps_in_events(array $in_events): array
 {
     $out_events = [];
 
     // Creation at the start
     if (count($in_events) == 0 || $in_events[0]['event_type'] != 'creation') {
         $pseudo_event = [
-            'timestamp' => '?',
```

Comment:
Background: `project_events` was added sometime after the site was live and so not all projects have records in the table. This creates pseudo records and for unknown timestamps we set it to `?` instead and output that value. This special value is used up on [line 1196](https://github.com/DistributedProofreaders/dproofreaders/pull/1346/files#diff-7b684f769c53e42bad88ffd69e373b9df6082ae4a6410d661aaf653a3865880eR1196) to output this value. For an example of this on PROD, see [projectID40bfb08353e74](https://www.pgdp.net/c/project.php?id=projectID40bfb08353e74).

If we want to make this a little more type friendly we could make `timestamp` an int-or-null and set it to null here and update line 1196.

Reply:
It would make more sense to me if this function pulled from the database, filled in any gaps, and return the final set instead of having the caller do the query and then pass it into here. Not that it needs addressing in this PR.

Reply:
Thanks for the context, that's very useful.

> If we want to make this a little more type friendly we could make timestamp an int-or-null and set it to null here and update line 1196.

I have made this change. An unfortunate side effect is that we allow `null` which will never be returned from the DB (the field is marked as `NOT NULL default 0`). This seemed like minor diff though.